File: services/project/internal/data/models/task.go
package models

import (
	"google.golang.org/protobuf/types/known/timestamppb"
	"pms.pkg/consts"
	"pms.pkg/transport/grpc/dto"
	"pms.pkg/type/timestamp"
	"pms.pkg/utils"
)

type Task struct {
	ID        string              `db:"id"`
	Title     string              `db:"title"`
	Body      string              `db:"body"`
	Status    consts.TaskStatus   `db:"status"`
	Priority  *int                `db:"priority"`
	ProjectID string              `db:"project_id"`
	SprintID  *string             `db:"sprint_id"`
	DueDate   timestamp.Timestamp `db:"due_date"`
	Created   timestamp.Timestamp `db:"created_at"`
	Updated   timestamp.Timestamp `db:"updated_at"`
}

func (t Task) DTO() *dto.Task {
	return &dto.Task{
		Id:        t.ID,
		Title:     t.Title,
		Body:      t.Body,
		Status:    string(t.Status),
		SprintId:  utils.Value(t.SprintID),
		ProjectId: t.ProjectID,
		Priority:  int32(utils.Value(t.Priority)),
		CreatedAt: timestamppb.New(t.Created.Time),
		UpdatedAt: timestamppb.New(t.Updated.Time),
		DueDate:   timestamppb.New(t.DueDate.Time),
	}
}

File: services/frontend/src/lib/task/task.ts
import { DeepPartial } from "../utils/generics";
import { Pagination } from "../utils/list";
import { TaskType } from "./tasktype";

export type Task = {
  id: string;
  title: string;
  body: string;
  status: string;
  task_type?: TaskType;
  type?: TaskType;
  sprint_id?: string;
  project_id?: string;
  priority: number;
  code: string;
  created_at: {
    seconds: number;
  };
  updated_at: {
    seconds: number;
  };
  assignee_id?: string;
  due_date: {
    seconds: number;
  };
};

export type TaskCreation = Omit<
  Task,
  "id" | "created_at" | "updated_at" | "code"
>;

export type TaskOptional = DeepPartial<Task>;

export type TaskFilter = Pagination & {
  sprint_id?: string;
  sprint_name?: string;
  status?: string;
  type?: TaskType;
  assignee_id?: string;
  priority?: number;
  title?: string;
  project_id?: string;
  project_name?: string;
  code?: string;
};
